docs/basics/architecture.md

```diff
@@ -0,0 +1,127 @@
+---
+title: Architecture
+order: 1
+---
+
+It may be useful to understand how Wonka's sources work internally
+if you want to write a new operator from scratch or contribute to it.
+
+This section explains how Wonka works internally and how it differs from
+the callbag specification.
+
+## Just Functions
+
+Internally Wonka only uses functions with rather simple signatures to
+make its streams work.
+
+We have sinks on one end, which need to receive values, and sources
+on the other, which need to send values.
+The sink is therefore just a function that we call with values over time.
+This is called a "push" signal.
+
+Because a sink has a start, incoming values, and an end, there are three
+signals that a sink can receive: `Start`, `Push`, and `End`.
+
+``` reason
+type signalT('a) =
+  | Start
+  | Push('a)
+  | End;
+
+type sinkT('a) = (. signalT('a)) => unit;
+```
+
+So the sink is just a function of `sinkT('a)`, which accepts a signal. When
```

Comment:
It may read nicer to split these `when`s into separate sentences.

```
When the stream starts, the sink is called with the `Start` signal. When a new value
comes in, the sink is called with `Push('a)` with the streamed value of type `'a`. Finally, when the stream ends
the sink is called with `End`.
```